lib/matplotlib/axes/_axes.py

```diff
@@ -978,10 +982,17 @@ def axhspan(self, ymin, ymax, xmin=0, xmax=1, **kwargs):
         self._check_no_units([xmin, xmax], ['xmin', 'xmax'])
         (ymin, ymax), = self._process_unit_info([("y", [ymin, ymax])], kwargs)
 
-        verts = (xmin, ymin), (xmin, ymax), (xmax, ymax), (xmax, ymin)
-        p = mpatches.Polygon(verts, **kwargs)
+        p = mpatches.Rectangle((xmin, ymin), xmax - xmin, ymax - ymin, **kwargs)
         p.set_transform(self.get_yaxis_transform(which="grid"))
+        # For Rectangles and non-separable transforms, add_patch can be buggy
+        # and update the x limits even though it shouldn't do so for an
+        # yaxis_transformed patch, so undo that update.
```

Comment:
There is some cutout in `_update_patch_limits` for x/y axis transforms; is that something that should be changed instead?

Reply:
I'll look into that.

Reply:
The problem is that Rectangle.get_transform() is not equal to get_xaxis/yaxis_transform, because Rectangle is implemented as a unit rectangle plus a sizing and translation affine transform, which ends up being included (first!) in Rectangle.get_transform.  I guess ideally we'd want to check something like `(-patch.get_patch_transform() + patch.get_transform())` (the subtraction needs to be "on the left"), but the transform system can't even handle that, so for now I don't think that would be so easy.